src/main/flight/autopilot.c

```diff
@@ -91,14 +155,209 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+#endif
+        ) {
+        return false;
+        // If compass is healthy, we must have a Mag, and therefore are OK, even with no GPS Heading
+        // If user allows posHold without a Mag, the IMU must be able to get heading from the GPS
+    }
+
+    // collect initial data values - gpsSol.llh = current gps location
+    int32_t bearing = 0; // degrees * 100
+    const float gpsDataIntervalS = getGpsDataIntervalSeconds(); // interval for current GPS data value 0.01s to 1.0s
+    const float gpsDataIntervalHz = 1.0f / gpsDataIntervalS;
+
+    // get distance and bearing from current location (gpsSol.llh) to target location
+    GPS_distance_cm_bearing(&gpsSol.llh, &currentTargetLocation, false, &posHold.distanceCm, &bearing);
+
+    // at the start, if the quad was moving, it will initially show increasing distance from start point
+    // once it has 'stopped' the PIDs will push back towards home, and the distance away will decrease
+    // it looks a lot better if we reset the target point to the point that we 'pull up' at
+    // otherwise there is a big distance to pull back if we start pos hold while carrying some speed
+    if (posHold.justStarted) {
+        if (posHold.distanceCm < posHold.previousDistanceCm) {
+            currentTargetLocation = gpsSol.llh;
+            resetPositionControlParams();
+            posHold.justStarted = false;
+        } else {
+            posHold.previousDistanceCm = posHold.distanceCm;
+        }
+    }
+    const uint8_t startLogger = posHold.justStarted ? 2 : 1;;
+    DEBUG_SET(DEBUG_AUTOPILOT_POSITION, 3, startLogger);
+    
+    // simple (very simple) sanity check
+    // primarily to detect flyaway from no Mag or badly oriented Mag
+    // TODO - maybe figure how to make a better check by giving more leeway at the start?
+    if (posHold.distanceCm > 1000) {
+        return false; // must stay within 10m or probably flying away
+        // value at this point is a 'best guess' to detect IMU failure in the event the user has no Mag
+        // if entering poshold from a stable hover, we would only exceed this if IMU was disoriented
+        // if entering poshold at speed, it may overshoot this value and falsely fail, if so need something more complex
+    }
+
+    // calculate error angle and normalise range
+    float errorAngle = (attitude.values.yaw * 10.0f - bearing) / 100.0f;
+    float normalisedErrorAngle = fmodf(errorAngle + 360.0f, 360.0f);
+    if (normalisedErrorAngle > 180.0f) {
+        normalisedErrorAngle -= 360.0f; // Range: -180 to 180
+    }
+
+    // Calculate distance error proportions for pitch and roll
+    const float errorAngleRadians = normalisedErrorAngle * (M_PIf / 180.0f);
+    const float rollProportion = -sin_approx(errorAngleRadians); // + 1 when target is left, -1 when to right, of the craft
+    const float pitchProportion = cos_approx(errorAngleRadians); // + 1 when target is ahead, -1 when behind, the craft
+
+    // set the filter gain used for D and J
```

Comment:
Jerk, as in 3D printer terminology.
It's really 'acceleration`.
So we could use `PIDA` or `PIDJ` - I thought that @KarateBrot preferred `J` for `Jerk` to mean acceleration, since that's a common 3D printer description of this kind of parameter.

But I would personally be happy with `A` for acceleration.

Particularly as, in theory, our filtered accelerometer data could be used to provide this value; or a crossfaded / Kalman acceleration value could be used that combined both sources.

@KarateBrot @mituritsyn what are your thoughts?  A or J?

Reply:
Vote for `A`

Reply:
Jerk is the derivative of acceleration. In 3D printer terminology this is also the derivative of acceleration. Jerk and acceleration are different quantities and not to be mistaken. Acceleration is usually abbreviated with "a", so the correct one is `A`

Reply:
Thanks @haslinghuis ...  I'll make it `A`, since it is 'acceleration' of velocity.  Makes sense.

I added acceleration for two reasons:
- first, the response to an attitude change is very delayed, with a non-zero angle acting to accelerate the quad, with some lag before a distance change happens, then it happens quickly.  So we need a very responsive PID system that focuses a lot on lag.
- and the derivative is quite noisy, and the filter causes lag, which is not helpful
- Acceleration has to be filtered even more, and that delays it, and also provides a kind of integration.  So the resulting filtered acceleration is sort of half acceleration and half velocity.
- Since fundamentally we are targeting zero velocity, acceleration is, in a sense, the derivative of what we are trying to control, and distance P is a bit more like I if we think of the controller as a velocity controller.  
- It practice, it works; in the logs, it helps make the response to uncommanded deflections quite a lot quicker.

This shows the PIDs where there is a quick direction error change.  The Acceleration factor helps D respond more quickly and overcomes some filter lag.
<img width="1218" alt="Screen Shot 2024-10-18 at 10 44 59 pm" src="https://github.com/user-attachments/assets/38887aef-0609-40d3-a869-4115d76acbaf">